test/extended/apiserver/tls.go

```diff
@@ -1,31 +1,523 @@
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		err = ForwardPortsAndExecute(
+			*restConfig,
+			&pods.Items[0],
+			[]string{"2379"},
+			3,
+			200*time.Millisecond,
+			func() {
+				// We aren't actually going through mTLS authentication with etcd to communicate
+				// with it - just checking TLS protocol versions. So, if it throws a "bad certificate"
+				// error, we're past the version check and consider it a success for this test.
+
+				conn, err := tls.Dial("tcp", "localhost:2379", tlsShouldWork)
+				if err != nil {
+					o.Expect(err.Error()).To(o.ContainSubstring("remote error: tls: bad certificate"))
+				} else {
+					conn.Close()
+				}
+
+				_, err = tls.Dial("tcp", "localhost:2379", tlsShouldNotWork)
+				o.Expect(err).To(o.HaveOccurred())
+			},
+		)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+	})
+})
+
+func ForwardPortsAndExecute(restConfig rest.Config, pod *v1.Pod, ports []string, maxConnectRetries int, initialBackoff time.Duration, toExecute func()) error {
+	if len(ports) < 1 {
+		return fmt.Errorf("at least 1 PORT is required for port-forward")
+	}
+
+	restClient, err := rest.RESTClientFor(defaultRESTConfigs(restConfig))
+	if err != nil {
+		return err
+	}
+
+	stdout := bytes.NewBuffer(nil)
+	req := restClient.Post().
+		Resource("pods").
+		Namespace(pod.Namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	transport, upgrader, err := spdy.RoundTripperFor(&restConfig)
+	if err != nil {
+		return err
+	}
+
+	stopChannel := make(chan struct{})
+	readyChannel := make(chan struct{})
+	errorChannel := make(chan error, 1)
+	defer close(stopChannel)
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+	fw, err := portforward.New(dialer, ports, stopChannel, readyChannel, stdout, stdout)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		var err error
+
+		for attempt := 0; attempt < maxConnectRetries; attempt++ {
+			err = fw.ForwardPorts()
+			if err == nil {
+				break
+			}
+
+			if attempt < maxConnectRetries-1 {
+				backoff := initialBackoff * time.Duration(math.Pow(2, float64(attempt)))
+				time.Sleep(backoff)
+			}
+		}
+
+		if err != nil {
+			errorChannel <- err
+			close(stopChannel)
+		}
+	}()
+
+	select {
+	case <-readyChannel:
+		toExecute()
+	case err := <-errorChannel:
+		return fmt.Errorf("port forwarding failed: %w", err)
+	case <-time.After(30 * time.Second):
+		return fmt.Errorf("timed out waiting for port forwarding to be ready")
+	}
+
+	return nil
+}
+
+func defaultRESTConfigs(config rest.Config) *rest.Config {
+	if config.GroupVersion == nil {
+		config.GroupVersion = &schema.GroupVersion{Group: "", Version: "v1"}
+	}
+	if config.NegotiatedSerializer == nil {
+		config.NegotiatedSerializer = scheme.Codecs
+	}
+	if len(config.UserAgent) == 0 {
+		config.UserAgent = rest.DefaultKubernetesUserAgent()
+	}
+	config.APIPath = "/api"
+	return &config
+}
+
+func getIPFamilyForCluster(client exutil.CLI, namespace string) IPFamily {
```

Comment:
Is the code related to getting the current IP family exactly the same as the one used in the existing network tests? 

If so, maybe we could move it to a common place and reuse it?

Reply:
ping ^ (we could open a new PR to refactor)

Reply:
It's mostly the same, but it adds a security context to let the check pod run in restricted contexts

Reply:
ping